src/dataflow/src/sink/kafka.rs

```diff
@@ -394,10 +397,40 @@ impl KafkaSinkState {
         }
     }
 
-    fn send<K, P>(&self, record: BaseRecord<K, P>) -> Result<(), bool>
+    async fn init_transactions(&self, timeout: Duration) -> KafkaResult<()> {
+        let self_producer = self.producer.clone();
+        task::spawn_blocking(move || self_producer.init_transactions(timeout))
+            .await
+            .unwrap_or(Err(KafkaError::Canceled))
```

Comment:
Maybe that's not possible, but could we forward the real error from the Kafka call?

Reply:
The error is forwarded, this `unwrap_or` is there because `spawn_blocking` wraps the result of the function in its own [`Result<T, JoinError>`](https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#associatedtype.Output) so in this case it's `Result<Result<(), KafkaError>>`.

`spawn_blocking` does so that if the function you passed panicked, it catches the unwind and converts it in the `Err(..)` case.

Since you commented here, if we want to preserve the semantics of the code before it was asyncified we should do a plain `unwrap()` on the join handle and let the whole thing panic as it would before.

Reply:
Gotcha! I think either the new or old behaviour is fine then.